plugin/formatting.py

```diff
@@ -100,16 +102,70 @@ class LspFormatDocumentCommand(LspTextCommand):
 
-    def run(self, edit: sublime.Edit, event: Optional[dict] = None) -> None:
-        format_document(self).then(self.on_result)
+    def run(self, edit: sublime.Edit, event: Optional[dict] = None, select: bool = False) -> None:
+        session_names = [session.config.name for session in self.sessions(self.capability)]
+        base_scope = self.view.scope_name(0).split()[0]
+        if select:
+            self.select_formatter(base_scope, session_names)
+        elif len(session_names) > 1:
+            window = self.view.window()
+            if not window:
+                return
+            formatter = None
+            project_data = window.project_data()
+            if isinstance(project_data, dict):
+                formatter = project_data.get('settings', {}).get('LSP', {}).get('formatters', {}).get(base_scope)
+            else:
+                window_manager = windows.lookup(window)
+                if window_manager:
+                    formatter = window_manager.formatters.get(base_scope)
+            if formatter:
+                session = self.session_by_name(formatter, self.capability)
+                if session:
+                    session.send_request_task(text_document_formatting(self.view)).then(self.on_result)
+                    return
+            self.select_formatter(base_scope, session_names)
+        else:
+            format_document(self).then(self.on_result)
 
     def on_result(self, result: FormatResponse) -> None:
         if result and not isinstance(result, Error):
             apply_text_edits_to_view(result, self.view)
 
+    def select_formatter(self, base_scope: str, session_names: List[str]) -> None:
+        window = self.view.window()
+        if not window:
+            return
+        window.show_quick_panel(
+            session_names, partial(self.on_select_formatter, base_scope, session_names), placeholder="Select Formatter")
+
+    def on_select_formatter(self, base_scope: str, session_names: List[str], index: int) -> None:
+        if index > -1:
+            session_name = session_names[index]
+            window = self.view.window()
+            if window:
+                window_manager = windows.lookup(window)
+                if window_manager:
+                    project_data = window.project_data()
+                    if isinstance(project_data, dict):
+                        project_settings = project_data.setdefault('settings', dict())
+                        project_lsp_settings = project_settings.setdefault('LSP', dict())
+                        project_formatter_settings = project_lsp_settings.setdefault('formatters', dict())
+                        project_formatter_settings[base_scope] = session_name
+                        # Prevent restart of all sessions after project file save
+                        window_manager.formatter_updated_in_project = True
+                        window.set_project_data(project_data)
```

Comment:
DottedDict should also help to make this shorter here?

Reply:
Can it? You need to still store the `project_data` as a normal Python dict for `set_project_data`, and DottedDict doesn't have any method to export to dict, because it would be ambiguous.

I just copied this code from https://github.com/sublimelsp/LSP/blob/fbd9652438e0b3b71362493ad6d9c6c2b9dce55f/plugin/core/workspace.py#L141-L146

Reply:
`DottedDict.get()` without argument will return the internal dict. There is also a `DottedDict.copy()` that will return a copy.

Reply:
I see. But it still won't work because we can't set the base scopes as entire keys in the DottedDict, because they contain a dot in the scope name. If we would try that, they would be splitted like
```json
{
  "formatters": {
    "source": {
      "js": "LSP-typescript"
    }      
  }
}
```
when saving in the project file.